src/diem/testing/miniwallet/app/offchain_api_v2.py

```diff
@@ -102,6 +116,43 @@ def _handle_offchain_payment_command(self, sender_address: str, request: Command
             account_id = self.app.store.find(Subaddress, subaddress_hex=subaddress).account_id
             self._create_payment_command(account_id, new_offchain_cmd, validate=True)
 
+    def _handle_offchain_reference_i_d_command(
+        self, request_sender_address: str, request: offchain.CommandRequestObject
+    ) -> Dict[str, Any]:
+        if request.command_type != CommandType.ReferenceIDCommand:
+            raise protocol_error(
+                ErrorCode.unknown_command_type,
+                f"unknown command_type: {request.command_type}",
+                field="command_type",
+            )
+        ref_id_command_object = from_dict(request.command, ReferenceIDCommandObject)
+
+        # Check if reference ID is duplicate
+        try:
+            self.app.validate_unique_reference_id(ref_id_command_object.reference_id)
+            # Check if receiver has a diem ID in this wallet
+            try:
+                account = self.app.store.find(Account, diem_id=ref_id_command_object.receiver)
+            except NotFoundError:
+                raise command_error(
+                    ErrorCode.invalid_receiver,
+                    f"Receiver with Diem ID {ref_id_command_object.receiver} not found",
+                    field="receiver",
+                )
+            self.app.store.create(
+                ReferenceID,
+                account_id=account.id,
+                reference_id=ref_id_command_object.reference_id,
+            )
+            return to_dict(
+                ReferenceIDCommandResultObject(
+                    receiver_address=self.app.diem_account.account_identifier(),
+                )
+            )
+        except ValueError:
```

Comment:
Is this error from validate_uniq reference id? If that's the case, should catch at the place the validation method is called. ValueError is a very common error, not sure whether there is another problem raises same error type and it may cause confusing error message